File: scripts/node/parsers/genericParser.js
'use strict';

const fs = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

function deriveApiName(fileName) {
  return fileName.replace(/\.[a-zA-Z]+-meta\.xml$/, '');
}

/**
 * Fallback parser used for any metadata type without a dedicated parser.
 * Walks every *-meta.xml file under typeDir (recursively, for nested-folder
 * types) and extracts whatever common fields are present.
 *
 * @param {string} typeDir - path to a metadata type folder (e.g. src/dashboards)
 * @returns {{ items: Array }}
 */
async function parseGeneric(typeDir) {
  const items = [];
  if (!fs.existsSync(typeDir)) return { items };

  const walk = (dir) => {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        walk(full);
      } else if (entry.name.endsWith('-meta.xml')) {
        const xml = fs.readFileSync(full, 'utf8');
        items.push({
          apiName: deriveApiName(entry.name),
          label: extractTag(xml, 'label') || extractTag(xml, 'masterLabel'),
          fullName: extractTag(xml, 'fullName'),
          description: extractTag(xml, 'description'),
        });
      }
    }
  };

  walk(typeDir);
  return { items };
}

module.exports = { parseGeneric };

File: scripts/python/parsers/generic_parser.py
"""
generic_parser.py
Fallback parser for any metadata type without a dedicated parser.
Extracts apiName/label/fullName/description from *-meta.xml files.
"""

import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def _derive_api_name(file_name: str) -> str:
    return re.sub(r'\.[a-zA-Z]+-meta\.xml$', '', file_name)


def parse_generic(type_dir: str) -> dict:
    """Walk every *-meta.xml file under type_dir and extract common fields."""
    items = []
    root = Path(type_dir)
    if not root.exists():
        return {'items': items}

    for file in sorted(root.rglob('*-meta.xml')):
        xml = file.read_text(encoding='utf-8')
        items.append({
            'apiName': _derive_api_name(file.name),
            'label': _extract_tag(xml, 'label') or _extract_tag(xml, 'masterLabel'),
            'fullName': _extract_tag(xml, 'fullName'),
            'description': _extract_tag(xml, 'description'),
        })

    return {'items': items}
